### DISCUSSION TITLE
The license status of this project

### DISCUSSION THREAD

**Discussion by KOLANICH** [GitoxideLabs/gitoxide#253] [Q&A] [answered]
Hi. Thank you for the awesome permissively-licensed git implementation I can use in my permissively-licensed libs.

ButI have some doubts. Even if the impl is not derived from the code in https://github.com/git/git , docs there is also under GPL. Since `git` itself is the original work, anything compatible to `git` is derived from it (it does not not necessarily mean that your work becomes a `derivative work` for the purpose of GPL) in any way. `git` is licensed under GPL, it was and is a choice of `git` authors and maintainers.

People willfully choosing GPL are usually big fans of using copyright in malicious ways (I usually internally think of such people as "copyright trolls" (in fact I use another word, but it is cannot be fully correctly translated into English) ). GPL is itself designed to be a viral and monopolizing license, and people choosing it obviously want to exploit these properties of the license.

Here are some examples:

1. When I asked `git` devs to relicense the docs under a permissive license in order to ensure that alternative permissive impls are licensially clean without being involved into a lawsuit, I [was told](https://lore.kernel.org/git/xmqqeexbrffe.fsf@gitster-ct.c.googlers.com/) that J. H. is not convinced that he wants to help people considering `the necessary effort to comply with the license` ([black-box and clean-room reverse engineering is meant under it, the stuff invented to circumvent copyright on proprietary software in a legal way](https://lore.kernel.org/git/4B11ADEA-8E8F-444F-B4CA-C4A3E9536F8C@mail.ru/#r)) `as a waste of time`.

2. https://www.mercurial-scm.org/pipermail/mercurial/2011-March/037593.html gives some insight into a mind of one GPL fan. Not necessarily into mind of all GPL fans, but it feels very consistent with the spirit of GPL and positions of orgs surrounding it.

3. This feeling is reinforced by FSF FAQ in which the clarification is done to interpret GPL as wide as they meant it be interpreted and make as much FUD as possible.

So I wonder if this impl is 100% licensially clean using the most restrictive GPL interpretation possible (i.e. clean room / black box reverse engineered)?

One can argue that ideas are not copyrightable (but are de-facto patentable, even though it is usually said that ideas are unpatentable, essentially it is that ideas are patentable, one just has to write the correct "magic words") and that using "elements dictated by external factors" is (sometimes, absolutely not always, and it is sometimes has to be tested in courts) allowed, but it is not about ideas, it is about a license, which exploits copyright law to prohibit people to do certain things. I.e. some (in fact, most of proprietary ones, though I have seen some licenses explicitly allowing it, most notably IDA Pro EULA) licenses explicitly prohibit any reverse engineering in any way.

**Comment by Byron** [comment, marked as answer]
Thanks for the excursion into copyright and GPL! The answer comes first as TL;DR:

> So I wonder if this impl is 100% licensially clean using the most restrictive GPL interpretation possible (i.e. clean room / black box reverse engineered)?

No, this implementation is not reverse-engineering `git`. I read the documentation and I read the source code.

Under less restrictive interpretations of the GPL I believe it's safe to use `gitoxide`, and looking at other alternative implementations and their licenses I think there are precedents for alternative licenses as well.

Saying that _I am not a lawyer_ is probably advisable here and this is not legal advice. It's fun and a little bit frightening to have to write such thing here 😅.

----

More thoughts follow…

> People willfully choosing GPL are usually big fans of using copyright in malicious ways (I usually internally think of such people as "copyright trolls" (in fact I use another word, but it is cannot be fully correctly translated into English) ).

I found [this article](https://ipkitten.blogspot.com/2019/02/copyright-trolling-abusive-litigation.html) when looking for `gpl copyright trolls` just to see where this is coming from. Seeing this happened in Germany of all places is very interesting, as it seems injunctions are cheap and fast there which gives leverage when suing companies, making abuse possible and beneficial for the one suing. 

> People willfully choosing GPL are usually big fans of using copyright in malicious ways (I usually internally think of such people as "copyright trolls" (in fact I use another word, but it is cannot be fully correctly translated into English) ).

I don't think of the copyright holders of `git` as potential patent trolls because they are happy with the choice of license despite its flaws.

> When I asked git devs to relicense the docs under a permissive license in order to ensure that alternative permissive impls are licensially clean without being involved into a lawsuit, I was told that J. H. is not convinced that he wants to help people considering the necessary effort to comply with the license (black-box and clean-room reverse engineering is meant under it, the stuff invented to circumvent copyright on proprietary software in a legal way) as a waste of time.

The way I understand the `git`maintainers is that they don't believe there is a need to relicense the docs as they believe it's fair to implement your own software based on it no matter which license.

> https://www.mercurial-scm.org/pipermail/mercurial/2011-March/037593.html gives some insight into a mind of one GPL fan. Not necessarily into mind of all GPL fans, but it feels very consistent with the spirit of GPL and positions of orgs surrounding it.

What an interesting read which made clear to me what clean-room implementation would mean to avoid this exact interpretation of the GPL. It's great that the `git` maintainers don't seem to share this sentiment.

**Reply by KOLANICH** [reply]
Thanks for the clarification.

>Saying that I am not a lawyer is probably advisable here and this is not legal advice. It's fun and a little bit frightening to have to write such thing here 😅.

Thanks for catching that. That was not a law advice, am not a lawyer, and the most importantly, not anyones' lawyer/attorney :) (though in my jurisdiction it is legal to give law advice without a license). My understanding of so called "exclusive right" legislation is based solely on reading of the articles in the Internet.

>I found this article when looking for gpl copyright trolls

Thanks for the article. I haven't really mind that article (and wasn't even aware of it) and the sense of the phrase used in the article, for me a "copyright troll" is everyone who approves so called "exclusive right" system or maliciously (imposing duties (maybe except attribution (I consider it is moral to give a proper attribution, not attributing may mean (again, there are some legit reasons to strip this info) that a person is an asshole, but if someone is an asshole, it is his choice to be an asshole, I don't care if someone took my code, removed my name from it and placed an own name on it instead, it won't make him the creator of that code and he would just look stupid when someone discovered that fact) and carrying license text and undeleteable copyright notices with the work (again, there are legit reasons not to carry them, i.e. I don't think it is right to waste microcontrollers' resources on storing and processing copyright notices and messages and license texts), but I don't think it is moral to enforce these things using laws) on the licensees and taking their inherent rights) uses it.

**Reply by Byron** [reply]
> Thanks for the article. I haven't really mind that article [..] and the sense of the phrase used in the article, for me a "copyright troll" is everyone who approves so called "exclusive right" system or maliciously [..] uses it.

Exclusive right systems are something I believe may be more harmful than they are helpful, thinking instead that collaborative approaches are the way to go especially if long-term sustainability and improvement is a goal. Yes, this means spreading possible proceeds over a greater amount of people while loosing some control, but ideally the collaborative spirit is coming along with it so that most players will generally be helpful with each other.

And of course I do hope that `gitoxide` will be part of such spirit :).
